### ISSUE TITLE
Consider making Scores.scores a structured array

### ISSUE THREAD

**Issue by florian-huber** [matchms/matchms#143]
Since we are moving towards inheriting from a base class for the similarity scores, I think it could be nice to switch to structured arrays for ``Scores.scores``.

Based on #135 (or a test implementation in PR #139 ) we could change the base class to: 
```python
class BaseSimilarityFunction:
    """Similarity function base class.
    When building a custom similarity measure, inherit from this class and implement
    the desired methods.
    """
    # Set key characteristics as class attributes
    is_commutative = True
    # Set output data type, e.g. ("score", "float") or [("score", "float"), ("matches", "int")]
    score_datatype = ("score", "float")

    def pair(self, reference: SpectrumType, query: SpectrumType) -> float:
        """...
        """
        raise NotImplementedError

    def matrix(self, references: List[SpectrumType], queries: List[SpectrumType],
               is_symmetric: bool = False) -> numpy.ndarray:
        """...
        """
        n_rows = len(references)
        n_cols = len(queries)
        scores = numpy.empty([n_rows, n_cols], dtype=self.score_datatype)
        for i_ref, reference in enumerate(references[:n_rows]):
        ...
```

For ``CosineGreedy`` we would then have to add:
```python
class CosineGreedy(BaseSimilarityFunction):
    """... """
    score_datatype = [("score", "float"), ("matches", "int")]
```
For ``ParentmassMatch`` it would become:
```python
class ParentmassMatch(BaseSimilarityFunction):
    """... """
    score_datatype = ("score", "bool")
```

That has 2 advantages in my eyes:
- It makes clearer what ``Scores.scores`` actually contains (name and data type).
- It can even be used to do splits in cases where we have tuples, e.g. ``scores["score"]`` vs ``scores["matches"]``

**Comment by florian-huber** [comment]
We could also enforce it a bit more by
```python
class BaseSimilarityFunction:
    """Similarity function base class.
    When building a custom similarity measure, inherit from this class and implement
    the desired methods.
    """
    # Set key characteristics as class attributes
    is_commutative = True
    # Set output data type, e.g. ("score", "float") or [("score", "float"), ("matches", "int")]
    score_datatype = ("score", "float")

    def compute_similarity_pair(self, reference: SpectrumType, query: SpectrumType) -> float:
        raise NotImplementedError

    def compute_similarity_matrix(self, references: List[SpectrumType], queries: List[SpectrumType],
                                  is_symmetric: bool = False) -> numpy.ndarray:
        ... loop here ...
        return scores

    def pair(self, reference: SpectrumType, query: SpectrumType) -> float:
        score = self.compute_similarity_pair(reference, query)
        return numpy.array(score, dtype=self.score_datatype)

    def matrix(self, references: List[SpectrumType], queries: List[SpectrumType],
               is_symmetric: bool = False) -> numpy.ndarray:
        scores = self.compute_similarity_matrix(references, queries)
        return numpy.array(scores, dtype=self.score_datatype)
        """...
```

**Comment by florian-huber** [comment]
Just tested it.
Generally works fine, but will brake our current main code example:
```python
import numpy as np
from matchms import calculate_scores
from matchms import Spectrum
from matchms.similarity import CosineGreedy

spectrum_1 = Spectrum(mz=np.array([100, 150, 200.]),
                      intensities=np.array([0.7, 0.2, 0.1]),
                      metadata={'id': 'spectrum1'})
spectrum_2 = Spectrum(mz=np.array([100, 140, 190.]),
                      intensities=np.array([0.4, 0.2, 0.1]),
                      metadata={'id': 'spectrum2'})
spectrum_3 = Spectrum(mz=np.array([110, 140, 195.]),
                      intensities=np.array([0.6, 0.2, 0.1]),
                      metadata={'id': 'spectrum3'})
spectrum_4 = Spectrum(mz=np.array([100, 150, 200.]),
                      intensities=np.array([0.6, 0.1, 0.6]),
                      metadata={'id': 'spectrum4'})
references = [spectrum_1, spectrum_2]
queries = [spectrum_3, spectrum_4]

similarity_measure = CosineGreedy()
scores = calculate_scores(references, queries, similarity_measure)

for (reference, query, score, n_matching) in scores:
    print(f"Cosine score between {reference.get('id')} and {query.get('id')}" +
          f" is {score:.2f} with {n_matching} matched peaks")
```
This will then expect something like:
```python
...
for (reference, query, (score, n_matching)) in scores:
    print(f"Cosine score between {reference.get('id')} and {query.get('id')}" +
          f" is {score:.2f} with {n_matching} matched peaks")
```

or
```python
...
for (reference, query, score) in scores:
    print(f"Cosine score between {reference.get('id')} and {query.get('id')}" +
          f" is {score["score"]:.2f} with {score["matches"]} matched peaks")
```

**Comment by florian-huber** [comment]
@sverhoeven I now made the underlying scores a structures array in PR #153, which I thought was a good move before addressing the additional querying options we discussed in #152 . 
However, it does change quiet some things. While it makes some things easier (e.g. calling `.scores["score"]` or being able to avoid some loops in the `Scores` methods), it might also make some things more complicated, for instance some tests where we now also have to check `dtype`. But also the similarity functions now need to specify the numpy array `dtype`.

What would you say: Move on to structured arrays or better leave it?

**Comment by florian-huber** [comment]
I have continued implementing it because it felt really comfortable be able to access the matrix of similarity scores by calling
```python
similarity_measure = CosineGreedy(tolerance=0.005)
similarity_matrix = similarity_measure.matrix(references, queries)

# just an example to get cosine scores with >= 6 matches
scores_selected = similarity_matrix["scores"][similarity_matrix["matches"] >= 6]  
```

**Comment by florian-huber** [comment]
Will be done with PR #153

**Comment by florian-huber** [comment]
Has now been merged with #153 .
